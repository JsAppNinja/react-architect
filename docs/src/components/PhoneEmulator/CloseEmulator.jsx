import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Button from 'react-md/lib/Buttons/Button';

export default class CloseEmulator extends PureComponent {
  static propTypes = {
    onClick: PropTypes.func,
  };

  static contextTypes = {
    hideDemo: PropTypes.func.isRequired,
  };

  handleClick = (e) => {
    if (this.props.onClick) {
      this.props.onClick(e);
    }

    this.context.hideDemo(e);
  };

  render() {
    return <Button {...this.props} onClick={this.handleClick} />;
  }
}
