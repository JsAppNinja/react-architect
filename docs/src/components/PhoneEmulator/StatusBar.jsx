import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import FontIcon from 'react-md/lib/FontIcons';

const StatusBar = ({ className, ...props }) => (
  <header {...props} className={cn('phone-emulator__status-bar', className)}>
    <FontIcon>network_wifi</FontIcon>
    <FontIcon>network_cell</FontIcon>
    <FontIcon>battery_full</FontIcon>
    <span className="phone-emulator__status-bar__time">12:30</span>
  </header>
);

StatusBar.propTypes = {
  className: PropTypes.string,
};
export default StatusBar;
