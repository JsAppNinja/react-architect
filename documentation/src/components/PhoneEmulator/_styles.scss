@import '~globals';

// scss-lint:disable NestingDepth
.phone-emulator {
  height: 100%;

  $gutter: $md-grid-phone-gutter / 2;

  &__content {
    height: 100%;
    padding-top: $md-toolbar-mobile-portrait-height;

    &.md-grid {
      padding-top: $md-toolbar-mobile-portrait-height + $gutter;
    }
  }

  @media #{$md-tablet-media} {
    @include react-md-toolbars-mobile;

    margin-left: auto;
    margin-right: auto;
    overflow: hidden;
    position: relative;
    width: $phone-emulator-width;

    .md-btn--fixed,
    .md-toolbar--fixed,
    .md-drawer--fixed,
    .md-snackbar,
    .md-overlay,
    .md-bottom-navigation {
      position: absolute;
    }

    .md-drawer--right {
      width: 100%;
    }

    .md-overlay {
      z-index: 1;
    }

    .md-drawer--right,
    .md-toolbar--fixed {
      z-index: $phone-emulator-toolbar-z-index;
    }

    .md-drawer--right,
    &__status-bar ~ .md-toolbar--fixed {
      top: $phone-emulator-status-bar-height;
    }

    .md-toolbar--prominent {
      height: $md-toolbar-prominent-height;
    }


    &__content {
      $height: $phone-emulator-status-bar-height;

      height: $phone-emulator-height;
      overflow: auto;
      padding-top: $height;

      &.md-grid {
        padding-top: $height + $gutter;
      }

      &--offset {
        $height: $height + $md-toolbar-mobile-portrait-height;
        padding-top: $height;

        &.md-grid {
          padding-top: $height + $gutter;
        }
      }

      &--offset-prominent {
        $height: $phone-emulator-status-bar-height + $md-toolbar-prominent-height;
        padding-top: $height;

        &.md-grid {
          padding-top: $height + $gutter;
        }
      }
    }

    &__toolbar {
      &.md-toolbar--fixed {
        top: $phone-emulator-status-bar-height;
      }

      &:not(.md-toolbar--inset)::before {
        background: $md-primary-color;
        content: '';
        height: $phone-emulator-status-bar-height;
        left: 0;
        position: absolute;
        right: 0;
        top: -$phone-emulator-status-bar-height;
      }
    }
  }

  @media #{$md-desktop-media} {
    @include react-md-typography-mobile;
    @include react-md-tabs-mobile;
    @include react-md-text-fields-mobile;
    @include react-md-text-field-custom-sizes;
    @include react-md-lists-mobile;
    @include react-md-buttons-mobile;
    @include react-md-snackbars-mobile;

    .material-icons.md-icon {
      font-size: $md-font-icon-size;
    }

    .md-btn--icon {
      height: $md-btn-icon-size;
      padding: $md-btn-icon-padding;
      width: $md-btn-icon-size;
    }
  }

  &__status-bar {
    align-items: center;
    background: $phone-emulator-status-bar-color;
    display: flex;
    height: $phone-emulator-status-bar-height;
    justify-content: flex-end;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    z-index: $phone-emulator-toolbar-z-index + 1;

    &__time {
      color: $md-white-base;
      font-weight: $md-font-medium;
      padding-right: .5em;
    }

    .md-icon {
      color: get-color('secondary', false);
    }

    @media #{$md-tablet-media} {
      ~ .md-toolbar-relative--prominent {
        margin-top: $md-toolbar-prominent-height + $phone-emulator-status-bar-height;
      }
    }
  }
}
