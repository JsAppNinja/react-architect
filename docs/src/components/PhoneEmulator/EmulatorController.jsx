import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { DialogContainer, Button } from 'react-md';

import Markdown from 'components/Markdown';
import PhoneEmulator from './PhoneEmulator';

const demoDescription = `
The current example will need to be opened in a dialog to be viewed because
it requires the full screen on mobile devices to be viewed correctly.
`;

export default class EmulatorController extends PureComponent {
  static propTypes = {
    mobile: PropTypes.bool,
    demoId: PropTypes.string,
    demoLabel: PropTypes.string,
    demoDescription: PropTypes.string,
    demoIconClose: PropTypes.node,
    toggleLabel: PropTypes.node,
    dialogStyle: PropTypes.object,
    dialogClassName: PropTypes.string,
  };

  static defaultProps = {
    demoId: 'phone-emulator-demo',
    demoLabel: 'A full page phone example',
    toggleLabel: 'View Example',
    demoDescription,
  };

  static childContextTypes = {
    hideDemo: PropTypes.func.isRequired,
  };

  state = { visible: false };

  getChildContext() {
    return { hideDemo: this.hideDemo };
  }

  showDemo = () => {
    this.setState({ visible: true });
  };

  hideDemo = () => {
    this.setState({ visible: false });
  };

  render() {
    const { visible } = this.state;
    const {
      demoId,
      demoLabel,
      demoDescription,
      demoIconClose,
      toggleLabel,
      dialogStyle,
      dialogClassName,
      ...props
    } = this.props;

    if (!props.mobile) {
      return <PhoneEmulator {...props} />;
    }

    let description;
    if (demoDescription) {
      description = <Markdown markdown={demoDescription} />;
    }

    return (
      <section>
        {description}
        <Button raised onClick={this.showDemo}>{toggleLabel}</Button>
        <DialogContainer
          id={demoId}
          visible={visible}
          aria-label={demoLabel}
          dialogStyle={dialogStyle}
          dialogClassName={dialogClassName}
          fullPage
        >
          <PhoneEmulator {...props} toolbarNavIcon={props.toolbarNavIcon || 'keyboard_arrow_left'} />
        </DialogContainer>
      </section>
    );
  }
}
