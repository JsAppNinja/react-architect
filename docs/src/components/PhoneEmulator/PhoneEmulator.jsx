import React, { Children } from 'react';
import CSSTransitionGroup from 'react-transition-group/CSSTransitionGroup';
import PropTypes from 'prop-types';
import cn from 'classnames';
import Paper from 'react-md/lib/Papers';
import Toolbar from 'react-md/lib/Toolbars';

import './_styles.scss';
import StatusBar from './StatusBar';
import CloseEmulator from './CloseEmulator';

const PhoneEmulator = ({
  style,
  className,
  contentStyle,
  contentClassName,
  toolbarStyle,
  toolbarClassName,
  toolbar,
  toolbarTitle,
  toolbarActions,
  toolbarZDepth,
  toolbarProminent,
  toolbarProminentTitle,
  toolbarNavIcon,
  mobile,
  mobileOnly,
  transitionContent,
  contentId,
  contentComponent,
  contentTransitionName,
  contentTransitionEnterTimeout,
  contentTransitionLeaveTimeout,
  children,
  tabsDemo,
}) => {
  const contentProps = {
    id: contentId,
    style: contentStyle,
    className: contentClassName,
    component: contentComponent,
    transitionName: contentTransitionName,
    transitionEnter: transitionContent && !!contentTransitionEnterTimeout,
    transitionEnterTimeout: contentTransitionEnterTimeout,
    transitionLeave: transitionContent && !!contentTransitionLeaveTimeout,
    transitionLeaveTimeout: contentTransitionLeaveTimeout,
  };

  if (mobileOnly && !mobile) {
    // If the demo is for mobile devices only, attempt to display the single child. If there are
    // multiple children, wrap it in the content component.
    if (Children.count(children) > 1) {
      return (
        <CSSTransitionGroup {...contentProps}>
          {children}
        </CSSTransitionGroup>
      );
    } else if (contentStyle || contentClassName) {
      const child = Children.only(children);
      return React.cloneElement(child, {
        style: { ...contentStyle, ...child.props.style },
        className: cn(contentClassName, child.props.className),
      });
    }

    return children;
  }

  let emulatorToolbar;
  let emulatorStatusBar;
  if (toolbar) {
    emulatorToolbar = (
      <Toolbar
        key="toolbar"
        fixed
        colored
        style={toolbarStyle}
        className={cn('phone-emulator__toolbar', toolbarClassName)}
        nav={<CloseEmulator icon>{toolbarNavIcon}</CloseEmulator>}
        actions={toolbarActions}
        title={toolbarTitle}
        prominent={toolbarProminent}
        prominentTitle={toolbarProminentTitle}
        zDepth={toolbarZDepth}
      />
    );
  }

  if (!mobile) {
    emulatorStatusBar = <StatusBar key="status-bar" />;
  }

  return (
    <Paper style={style} className={cn('phone-emulator md-background', className)} zDepth={mobile ? 0 : undefined}>
      {emulatorStatusBar}
      {emulatorToolbar}
      <CSSTransitionGroup
        {...contentProps}
        className={cn('phone-emulator__content', {
          'phone-emulator__content--padded': !tabsDemo && !mobile,
          'phone-emulator__content--offset': toolbar,
          'phone-emulator__content--offset-prominent': toolbar && (toolbarProminent || toolbarProminentTitle),
        }, contentClassName)}
      >
        {children}
      </CSSTransitionGroup>
    </Paper>
  );
};

PhoneEmulator.propTypes = {
  style: PropTypes.object,
  className: PropTypes.string,
  contentStyle: PropTypes.object,
  contentClassName: PropTypes.string,
  toolbarStyle: PropTypes.object,
  toolbarClassName: PropTypes.string,
  contentComponent: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.string,
  ]).isRequired,
  transitionContent: PropTypes.bool,
  contentId: PropTypes.string,
  contentTransitionName: PropTypes.string,
  contentTransitionEnterTimeout: PropTypes.number,
  contentTransitionLeaveTimeout: PropTypes.number,

  /**
   * Boolean if a toolbar should appear in the phone emulator demo.
   */
  toolbar: PropTypes.bool,
  toolbarTitle: PropTypes.node,
  toolbarActions: Toolbar.propTypes.actions,
  toolbarZDepth: PropTypes.number,
  toolbarProminent: PropTypes.bool,
  toolbarProminentTitle: PropTypes.bool,
  toolbarNavIcon: PropTypes.node,

  /**
   * Boolean if the emulator should appear only on mobile devices. When this is enabled,
   * the children will be displayed as-is instead of in the emulator.
   */
  mobileOnly: PropTypes.bool,
  tabsDemo: PropTypes.bool,
  mobile: PropTypes.bool,
  children: PropTypes.node,
};

PhoneEmulator.defaultProps = {
  toolbarTitle: 'Title',
  toolbarNavIcon: 'menu',
  contentComponent: 'section',
  contentTransitionName: 'md-cross-fade',
  contentTransitionEnterTimeout: 300,
  contentTransitionLeaveTimeout: 0,
  toolbar: true,
};

export default PhoneEmulator;
