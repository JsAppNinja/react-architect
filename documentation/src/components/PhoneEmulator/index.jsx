import { connectAdvanced } from 'react-redux';
import shallowEqual from 'shallowequal';
import EmulatorController from './EmulatorController';

export default connectAdvanced(() => {
  let result;

  return (state, props) => {
    const { mobile, tablet, desktop } = state.media;
    const nextResult = { ...props, mobile, tablet, desktop };
    if (!shallowEqual(result, nextResult)) {
      result = nextResult;
    }

    return result;
  };
})(EmulatorController);
