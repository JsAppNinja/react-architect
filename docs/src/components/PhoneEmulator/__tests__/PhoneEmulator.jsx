/* eslint-env jest */
import React from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';

import PhoneEmulator from '../PhoneEmulator';

class ContextProvider extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  static childContextTypes = {
    hideDemo: PropTypes.func.isRequired,
  };

  getChildContext() {
    return { hideDemo: jest.fn() };
  }

  render() {
    return this.props.children;
  }
}

describe('PhoneEmulator', () => {
  it('should render correctly with the base required props', () => {
    const tree1 = renderer.create(
      <ContextProvider>
        <PhoneEmulator mobile={false}>
          <h2>This is some content that should appear in the phone emulator!</h2>
        </PhoneEmulator>
      </ContextProvider>
    ).toJSON();
    const tree2 = renderer.create(
      <ContextProvider>
        <PhoneEmulator mobile>
          <h2>This is some content that should appear in the phone emulator!</h2>
        </PhoneEmulator>
      </ContextProvider>
    ).toJSON();
    expect(tree1).toMatchSnapshot();
    expect(tree2).toMatchSnapshot();
  });

  it('should be able to render without a toolbar', () => {
    const tree = renderer.create(
      <ContextProvider>
        <PhoneEmulator toolbar={false}>
          <h2>This is some content that should appear in the phone emulator!</h2>
        </PhoneEmulator>
      </ContextProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render only the children if mobileOnly is provided and not mobile', () => {
    const tree1 = renderer.create(
      <ContextProvider>
        <PhoneEmulator mobileOnly mobile={false}>
          <h2>This is some content that should appear in the phone emulator!</h2>
        </PhoneEmulator>
      </ContextProvider>
    ).toJSON();
    const tree2 = renderer.create(
      <ContextProvider>
        <PhoneEmulator mobileOnly mobile={false}>
          <div>Line 1</div>
          <div>Line 2</div>
        </PhoneEmulator>
      </ContextProvider>
    ).toJSON();

    expect(tree1).toMatchSnapshot();
    expect(tree2).toMatchSnapshot();
  });
});
