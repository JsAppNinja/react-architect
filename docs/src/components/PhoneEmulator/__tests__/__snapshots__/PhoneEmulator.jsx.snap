// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneEmulator should be able to render without a toolbar 1`] = `
<div
  class="md-paper md-paper--1 phone-emulator md-background"
>
  <header
    class="phone-emulator__status-bar"
  >
    <i
      class="md-icon material-icons"
    >
      network_wifi
    </i>
    <i
      class="md-icon material-icons"
    >
      network_cell
    </i>
    <i
      class="md-icon material-icons"
    >
      battery_full
    </i>
    <span
      class="phone-emulator__status-bar__time"
    >
      12:30
    </span>
  </header>
  <section
    class="phone-emulator__content"
  >
    <h2>
      This is some content that should appear in the phone emulator!
    </h2>
  </section>
</div>
`;

exports[`PhoneEmulator should render correctly with the base required props 1`] = `
<div
  class="md-paper md-paper--1 phone-emulator md-background"
>
  <header
    class="phone-emulator__status-bar"
  >
    <i
      class="md-icon material-icons"
    >
      network_wifi
    </i>
    <i
      class="md-icon material-icons"
    >
      network_cell
    </i>
    <i
      class="md-icon material-icons"
    >
      battery_full
    </i>
    <span
      class="phone-emulator__status-bar__time"
    >
      12:30
    </span>
  </header>
  <header
    class="md-paper md-paper--2 md-toolbar md-background--primary md-toolbar--text-white md-toolbar--fixed phone-emulator__toolbar"
  >
    <button
      class="md-btn md-btn--icon md-pointer--hover md-inline-block md-btn--toolbar md-toolbar--action-left"
      type="button"
    >
      <div
        class="md-ink-container"
      />
      <i
        class="md-icon material-icons md-text--inherit"
      >
        menu
      </i>
    </button>
    <h2
      class="md-title md-title--toolbar"
    >
      Title
    </h2>
  </header>
  <section
    class="phone-emulator__content phone-emulator__content--offset"
  >
    <h2>
      This is some content that should appear in the phone emulator!
    </h2>
  </section>
</div>
`;

exports[`PhoneEmulator should render correctly with the base required props 2`] = `
<div
  class="md-paper md-paper--0 phone-emulator md-background"
>
  <header
    class="md-paper md-paper--2 md-toolbar md-background--primary md-toolbar--text-white md-toolbar--fixed phone-emulator__toolbar"
  >
    <button
      class="md-btn md-btn--icon md-pointer--hover md-inline-block md-btn--toolbar md-toolbar--action-left"
      type="button"
    >
      <div
        class="md-ink-container"
      />
      <i
        class="md-icon material-icons md-text--inherit"
      >
        menu
      </i>
    </button>
    <h2
      class="md-title md-title--toolbar"
    >
      Title
    </h2>
  </header>
  <section
    class="phone-emulator__content phone-emulator__content--offset"
  >
    <h2>
      This is some content that should appear in the phone emulator!
    </h2>
  </section>
</div>
`;

exports[`PhoneEmulator should render only the children if mobileOnly is provided and not mobile 1`] = `
<h2>
  This is some content that should appear in the phone emulator!
</h2>
`;

exports[`PhoneEmulator should render only the children if mobileOnly is provided and not mobile 2`] = `
<section>
  <div>
    Line 1
  </div>
  <div>
    Line 2
  </div>
</section>
`;
